File: app/ws_2_0/connection_agent.py

# Copyright 2016 The COMPEIT Consortium
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging

from flask import request
from flask.ext.login import login_required, current_user
from flask.ext.socketio import emit, join_room, leave_room

from .. import socketio, db
from ..models import User, StatisticEvent
from ..models import add_event

import kurento_agent
import chat_agent
import file_agent

logger = logging.getLogger(__name__)

# WS Socket communication

ws_clients = {}


def user_emit(user_id, event, data):
    if int(user_id) in ws_clients:
        ws_clients[int(user_id)]['/connection_agent'].base_emit(event, data)
    else:
        print ws_clients
        logger.info('User id '+user_id+' not among connected ws clients')


@socketio.on('connect', namespace='/connection_agent')
@login_required
def dashboard_connect():
    logger.debug('Got ws connect user id: %d with session id: %s' % (current_user.id, request.namespace.socket.sessid))

    ws_clients[int(current_user.id)] = request.namespace.socket
    emit('connected', {'user': {'id': current_user.id, 'name': current_user.name}})


@socketio.on('join_room', namespace='/connection_agent')
def room_join(message):
    roomId = int(message['room'])
    logger.debug("got join_room for " + str(roomId))

    logger.debug(request.namespace.socket.sessid)

    if current_user.activeRoomId and int(current_user.activeRoomId) != int(roomId):
        logger.debug("leaving room " + str(current_user.activeRoomId) + " to room " + str(roomId))
        join_room(current_user.activeRoomId)
        emit('user_left', {'user': current_user.to_dict()}, room=int(current_user.activeRoomId))
        add_event(StatisticEvent.ROOM_LEAVE, userId=current_user.id, roomId=roomId)
        leave_room(current_user.activeRoomId)

    emit('room_joined')

    emit('user_joined', {'user': current_user.to_dict()}, room=roomId)
    add_event(StatisticEvent.ROOM_JOIN, userId=current_user.id, roomId=roomId)

    join_room(roomId)
    if not current_user.activeRoomId or int(current_user.activeRoomId) != int(roomId):
        current_user.activeRoomId = roomId
        db.session.commit()

    active_users = User.query.filter_by(activeRoomId=roomId).all()

    for user in active_users:
        if current_user.id != user.id and user.id in ws_clients:
            user_emit(current_user.id, 'user_joined', {'user': user.to_dict()})


@socketio.on('leave_room', namespace='/connection_agent')
def room_leave(message):
    if current_user.activeRoomId:
        roomId = current_user.activeRoomId
        emit('user_left', {'user': current_user.to_dict()}, room=roomId)
        leave_room(roomId)
        add_event(StatisticEvent.ROOM_LEAVE, userId=current_user.id, roomId=roomId)

    current_user.activeRoomId = None
    db.session.commit()

    emit('room_left', {})


@socketio.on('disconnect', namespace='/connection_agent')
@login_required
def dashboard_disconnect():
    if request.namespace.socket.sessid == ws_clients[current_user.id].sessid:
        if current_user.activeRoomId:
            emit('user_left', {'user': current_user.to_dict()}, room=int(current_user.activeRoomId))
            add_event(StatisticEvent.ROOM_LEAVE, userId=current_user.id, roomId=current_user.activeRoomId)

        current_user.activeRoomId = None
        db.session.commit()
        logger.debug('Client disconnected')
    else:
        logger.debug('Old session disconnected')


@socketio.on('server_message', namespace='/connection_agent')
def server_message(message):
    logger.debug('server_message(' + str(current_user.activeRoomId) + ') ' + str(message))

    if 'type' in message:
        if message['type'] == 'kurento':
            kurento_agent.server_message(message['data'])
        elif message['type'] == 'file':
            file_agent.server_message(message['data'])


@socketio.on('room_message', namespace='/connection_agent')
def room_message(message):
    logger.debug('room_message' + str(message) + ' (' + str(current_user.activeRoomId)+')')

    # Special message types

    if 'type' in message and message['type'] == 'chat':
        chat_agent.room_message(message)

    emit('room_message', {'user': current_user.to_dict(), 'type': message['type'], 'data': message['data']}, room=int(current_user.activeRoomId))


@socketio.on('user_message', namespace='/connection_agent')
def user_message(message):
    logger.debug('user_message' + str(message) + ' (' + str(current_user.activeRoomId)+')')

    if 'type' in message and message['type'] == 'chat':
        chat_agent.user_message(message)

    if 'to' in message and int(message['to']) in ws_clients:
        user_emit(message['to'], 'user_message', {'type': message['type'], 'data': message['data'], 'from': current_user.id})

File: app/static/js/connection_agent.js

/*
 * Copyright 2016 The COMPEIT Consortium
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
 
connectionAgent = function() {
  var socket = null;
  var eventCBs = [];
  var roomCBs = [];
  var userCBs = [];
  var roomId = 0;
  var me = null;
  var evtCB = null;


  function eventCB(evt, data) {
    for (var type in eventCBs) {
      eventCBs[type](evt, data);
    }
    if (evtCB) {
      evtCB(evt, data);
    }
  }

  function logTypes() {
    for (var t in eventCBs) {
      console.log(t);
    }
  }


  return {
    init: function(room) {
      roomId = room;
    },
    start: function() {
      var namespace = '/connection_agent'; // change to an empty string to use the global namespace

      // the socket.io documentation recommends sending an explicit package upon connection
      // this is specially important when using the global namespace
      //var socket = io.connect('http://' + document.domain + ':' + location.port + namespace);
      socket = io.connect(document.domain + ':' + location.port + namespace);

      socket.on('connect', function() {
        console.log('Connection established');
      });

      socket.on('connected', function(msg) {
        console.log('Connected to connection agent');
        me = msg['user'];
        eventCB.call(this, 'connected', msg['user']);
        socket.emit('join_room', {'room': roomId});
      });


      socket.on('room_joined', function(msg) {
        //console.log('room_joined');
        eventCB.call(this, 'room_joined', {})
      });

      socket.on('user_joined', function(msg) {
        console.log('user_joined');
        //console.log(msg);
        if (me['id'] != msg['user']['id']) {
          eventCB.call(this, 'user_joined', msg);
        }
      });

      socket.on('user_left', function(msg) {
        console.log('user_left');
        console.log(msg);
        if (me['id'] != msg['user']['id']) {
          eventCB.call(this, 'user_left', msg);
        }
      });

      socket.on('server_message', function(msg) {
        if ('type' in msg && msg['type'] in eventCBs) {
          if ('data' in msg) {
            eventCBs[msg['type']](msg['event'], msg['data']);
          } else {
            eventCBs[msg['type']](null, msg['data']);
          }
        }
      });

      socket.on('room_message', function(msg) {
        // console.log('room_message received');
        // console.log(msg);
        if ('type' in msg && msg['type'] in roomCBs) {
          if ('data' in msg) {
            roomCBs[msg['type']](msg['user'], msg['data']);
          } else {
            roomCBs[msg['type']](msg['user']);
          }
        }
      });

      socket.on('user_message', function(msg) {
        // console.log('user_message received');
        // console.log(msg);
        if ('type' in msg && msg['type'] in userCBs) {
          if ('data' in msg) {
            userCBs[msg['type']](msg['from'], msg['data']);
          } else {
            userCBs[msg['type']](msg['from']);
          }
        }
      });

    },
    stop: function() {
      socket.close();
      eventCB('stopped', {});
      socket = null;
    },
    registerType: function(type, callbackEvent, callbackRoom, callbackUser) {
      var myType = type;
      if (callbackEvent) {
        eventCBs[type] = callbackEvent;
      }
      if (callbackRoom) {
        roomCBs[type] = callbackRoom;
      }
      if (callbackUser) {
        userCBs[type] = callbackUser;
      }

      return {
        sendServerMessage: function (msg) {
          this.sendServerMessage(myType, msg);
        }.bind(this),
        sendRoomMessage: function (msg) {
          this.sendRoomMessage(myType, msg);
        }.bind(this),
        sendUserMessage: function(user_id, msg) {
          this.sendUserMessage(user_id, myType, msg);
        }.bind(this),
        getType: function() {
          return myType;
        }.bind(this),
        debug: function() {
          this.debug();
        }.bind(this)
      }
    },
    getName: function () {
      return me.name;
    },
    onEvent: function (cb) {
      evtCB = cb;
    },
    sendServerMessage: function(type, msg) {
      socket.emit('server_message', {'type': type, 'data': msg});
    },
    sendRoomMessage: function(type, msg) {
      socket.emit('room_message', {'room': roomId , 'type': type, 'data': msg});
    },
    sendUserMessage: function(user_id, type, msg) {
      // console.log('user_message sent '+user_id+' '+type+' '+JSON.stringify(msg));
      socket.emit('user_message', {'to': user_id, 'type': type, 'data': msg});
    },
    debug: function() {
      logTypes();
    }
  }
}();
